### DISCUSSION TITLE
Allow patching default Helm values

### DISCUSSION THREAD

**Discussion by janpieper** [mysticaltech/terraform-hcloud-kube-hetzner#1384] [Ideas]
## Problem

The default value for the `nginx_values` local is based on a ton of locals and variables:

https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/blob/dbcf426864cddb2c41d64e5a405918a648c98883/locals.tf#L546-L574

Same goes for [haproxy_values](https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/blob/dbcf426864cddb2c41d64e5a405918a648c98883/locals.tf#L576-L621) and [traefik_values](https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/blob/dbcf426864cddb2c41d64e5a405918a648c98883/locals.tf#L623-L738), etc.

To add a custom value, I am currently forced to copy the default value to my Terraform code, which could then cause inconsistencies when I change e.g. `load_balancer_algorithm_type` but forget to also update my custom `nginx_values`.

### Example

```hcl
module "kube-hetzner" {
  # ...

  load_balancer_algorithm_type = "least_connections"

  nginx_values = <<EOT
# ...
service:
  annotations:
    "load-balancer.hetzner.cloud/algorithm-type": "round_robin"
# ...
  EOT

  # ...
}
```

As you can see `load_balancer_algorithm_type` is set to `least_connections`, but the `load-balancer.hetzner.cloud/algorithm-type` annotation is set to `round_robin`.

To avoid this, one can add a local and use it for the module-variable and for the custom `nginx_value`:

```hcl
locals {
  load_balancer_algorithm_type = "least_connections"
}

module "kube-hetzner" {
  # ...

  load_balancer_algorithm_type = local.load_balancer_algorithm_type

  nginx_values = <<EOT
# ...
service:
  annotations:
    "load-balancer.hetzner.cloud/algorithm-type": "${local.load_balancer_algorithm_type}"
# ...
  EOT

  # ...
}
```

That works, but it's quite some work, because tons of variables are used and they also use some locals that are based on _calculations_:

https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/blob/dbcf426864cddb2c41d64e5a405918a648c98883/locals.tf#L230

https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/blob/dbcf426864cddb2c41d64e5a405918a648c98883/locals.tf#L233

### Idea

It would be cool to have a way to add custom options to the default values (e.g. some kind of merging strategy?) without being forced to copy the default value.

**Comment by mysticaltech** [comment]
@janpieper Makes me think of kustomizations, it uses that kind of merging strategies. Have a look and PRs are most welcome 🙏

**Comment by safizn** [comment]
A simple solution could be to add a different variable "*_values_merge" which if set would use Terraform's "merge" function. 

```
cilium_values = var.cilium_values != "" ? var.cilium_values : <<EOT
  ... 
EOT

cilium_values = merge( cilium_values, var.cilium_values_merge )
```

**Comment by mysticaltech** [comment]
This is done indeed, thanks for mentioning it @safizn!
